### DISCUSSION TITLE
PCF8574 Input double press functionality and rules simplification

### DISCUSSION THREAD

**Discussion by ChrisVonFrosta** [arendst/Tasmota#17836] [Q&A]
I have an esp32 with 4 PCF8574 expansion. each of 2 are for 8 Outputs/relays, the  other ones each are for 8 Inputs/Pushbutton.
So we have16 relays and 16 inputs in total.
As i figured Out the relays corresponding with the power numbers.

I know i can set Rules for each pair of Output and Input.
for example:
`Rule1 ON PCF8574-2_INP#D0=0 DO Power1 toggle ENDON`
`Rule2 ON PCF8574-2_INP#D1=0 DO Power2 toggle ENDON`
is the a smarter way to realize those rules for 16 pairs?

I want to realize a double press functionality to some or all inputs.
So i could switch 2 different relays with one push button.
for example:
single push to toggle relay1 and double push for relay10.
`SetOption11 1` seams doen't effect the ipnuts from PCF8574.
Is there an Option?
at long last i would even go with hardcoding.

Thanks for any help!

**Comment by barbudor** [comment]
As of today, external IO expanders do not support button features including multipress
This may change in the future as there are some internal works on going.
When the architecture is updated, it should be possible to upgrade the PCF driver to allow that.
For now it would require some major work.

**Comment by ChrisVonFrosta** [comment]
Thank you for this inside view.

whats is about the rules for all 16 channels/pairs?
Do i have to set 16 rules?
i tried something like this to get multiple channel done with one line:
`ON PCF8574-2_INP#D var1 %value% ENDON ON PCF8574-2_INP#D DO power%var1% ENDON`
`ON PCF8574-2_INP#D var1 %value% ENDON  DO power%var1% ENDON`

since it never worked, i tried rules like this for debugging (print var1 in console):

`ON PCF8574-2 DO var1 %value% ENDON ON PCF8574-2_INP DO var1 ENDON`
`ON PCF8574-2_INP DO var1 %value% ENDON ON PCF8574-2_INP DO var1 ENDON`
`ON PCF8574-2_INP#D DO var1 %value% ENDON ON PCF8574-2_INP DO var1 ENDON`

var1 returns with "1" allways.

Do i mistakes or is the event "PCF8574-X" not compatible for those requests?
THX!

**Reply by barbudor** [reply]
You need 1 rule per input like in your previous message

Don't mistake rule and ruleset
Rule1, Rule2, rule3 are ruleset which can have up to 512 characters and multiple rules

A rule is just 'on <trigger> do <command> endon' and you can have multiple per ruleset

**Reply by ChrisVonFrosta** [reply]
thx for additional explanation!
i tried to add 5 Rules (rule1 - rule5) and run in this mistake.

now i set only Rule1 with 8 pairs at frst. 
I hope the support for double press oder long press will coming soon.
I want to build this in my "smat home"

```
Rule1
ON PCF8574-2_INP#D0=0 DO Power1 toggle ENDON
ON PCF8574-2_INP#D1=0 DO Power2 toggle ENDON
ON PCF8574-2_INP#D2=0 DO Power3 toggle ENDON
ON PCF8574-2_INP#D3=0 DO Power4 toggle ENDON
ON PCF8574-2_INP#D4=0 DO Power5 toggle ENDON
ON PCF8574-2_INP#D5=0 DO Power6 toggle ENDON
ON PCF8574-2_INP#D6=0 DO Power7 toggle ENDON
ON PCF8574-2_INP#D7=0 DO Power8 toggle ENDON 
```

**Comment by shostkevych** [comment]
@barbudor , according to docs https://tasmota.github.io/docs/PCF8574/, there is special MODE_2 which is looking interesting in terms of mapping. Curious, if this setup will allow to achieve multipress and acting as usual buttons on PCF8574 ? 

I'm using this https://templates.blakadder.com/kincony_KC868-A8.html board and strongly interested in getting all the 8 inputs acting as standard tasmota button. 

Thanks in advance !

**Reply by barbudor** [reply]
Indeed, mode 2 use the new method that allows to map fully standard tasmota Buttons or Switches to extender pins with the same available features as on native GPIO.
There's no UI to set it up so you have to understand how the json configuration works

**Reply by oshostkevych-sparx** [reply]
@barbudor , thanks for the answer. Will look further into this.

**Reply by alqueres** [reply]
@barbudor , after one year is there any reference of how json configuration works? Are you aware of any use for an example?
Thanks in advance!

**Reply by sfromis** [reply]
Explained here, including examples:
https://tasmota.github.io/docs/PCF8574/#mode-2
